### PR TITLE
Object Ownership

### PR DESCRIPTION
**Pull request by cpcallen**
The final commit adds an .owner slot to (pseudo)Object.  Everything else is preparatory work.

I suggest you look at it commit-by-commit (and read commit messages) so you can decide which deserve closest scrutiny.

No tests yet, because without a getOwner function it's hard to decide if the test passed.

### CODE REVIEW

File: server/interpreter.js [hunk 1]
```diff
@@ -2459,22 +2461,21 @@ Interpreter.Thread.prototype.getSource = function(index) {
...
-
- * @return {!Interpreter.prototype.Object|undefined}
+ * user/group/etc. object).  It is an error to call this function on a
+ * thread that is a zombie.
+ * @return {?Interpreter.Owner}
  */
 Interpreter.Thread.prototype.perms = function() {
   if (this.status === Interpreter.Thread.Status.ZOMBIE) {
-    return undefined;
+    throw Error("Zombie thread has no perms");
```
Comment (NeilFraser) [inline server/interpreter.js]: Single quotes prefered.

File: server/interpreter.js [hunk 2]
```diff
@@ -1572,6 +1573,32 @@ Interpreter.prototype.initThreads_ = function() {
...
+  this.createNativeFunction('perms', function() {
+    if (!intrp.thread) {
+      throw Error('No current thread??');
+    }
+    return intrp.thread.perms();
+  }, false);
+
+  this.createNativeFunction('setPerms', function(perms) {
+    if (!intrp.thread) {
+      throw Error('No current thread??');
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: Fixed.

File: server/tests/db/test_02_perms.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+/**
+ * @license
+ * Code City: Testing code.
+ *
+ * Copyright 2017 Google Inc.
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/tests/db/test_02_perms.js]: Fixed; also in core_00_global.js, added in previous PR.

File: server/interpreter.js [hunk 4]
```diff
@@ -2429,6 +2456,33 @@ Interpreter.Thread.prototype.getSource = function(index) {
   return source;
 };
 
+/**
+ * Returns the permissions with which currently-executing code is
+ * running (equivalent to a unix EUID, but in the form of a
+ * user/group/etc. object) - or undefined if the thread is a zombie.e
+ * a zombie.
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: Dunno what happened here, but fixed already.

File: server/interpreter.js (hunk 1 above)
Comment (cpcallen) [reply server/interpreter.js]: Fixed.

**Pull request merged**